[AArch64][SME]Check streaming mode when using SME2 instruction in fra…

CarolineConcatto opened the pull request:
…me lowering

SME instructions can only be used in streaming mode. PTRUE for predicated counter and the ld/st pair can be used when:
  sve2.1  is available or
  sme2 available in function in streaming mode.
Previously the frame lowering only checking if sme2 available when building the machine instruction.
This fix checks if sme2 is available and is subtarget in streaming mode

CarolineConcatto committed:
[AArch64][SME]Check streaming mode when using SME2 instruction in frame lowering

SME instructions can only be used in streaming mode.
PTRUE for predicated counter and the ld/st pair can be used when:
  sve2.1  is available or
  sme2 available in function in streaming mode.
Previously the frame lowering only checking if sme2 available when building
the machine instruction.
This fix checks if sme2 is available and is subtarget in streaming mode

efriedma-quic commented:
Is the revised check correct with __arm_locally_streaming?

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2955,6 +2955,14 @@ unsigned findFreePredicateReg(BitVector &SavedRegs) {
 }
 
+bool isSve2p1OrSme2InStreaming(const AArch64Subtarget &Subtarget) {
+  if (Subtarget.hasSVE2p1())
+    return true;
+  if (Subtarget.hasSME2() && Subtarget.isStreaming())
+    return true;
+  return false;
```
What about:
```suggestion
return Subtarget.hasSVE2p1() || (Subtarget.hasSME2() && Subtarget.isStreaming())
```
?

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2955,6 +2955,14 @@ unsigned findFreePredicateReg(BitVector &SavedRegs) {
   return AArch64::NoRegister;
 }
 
+bool isSve2p1OrSme2InStreaming(const AArch64Subtarget &Subtarget) {
```
It would be nice to rename this to something like `bool enableMultiVectorSpillFill(..)` and add a command line option that allows users to disable this feature.

sdesmalen-arm commented on llvm/test/CodeGen/AArch64/sve-callee-save-restore-pairs.ll:
```diff
@@ -3,7 +3,8 @@
 ; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sve2p1 -verify-machineinstrs < %s | FileCheck %s --check-prefixes=PAIR
-
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sme2 -mattr=+sve2 -verify-machineinstrs < %s | FileCheck %s --check-prefixes=NOPAIR
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sme2  -mattr=+sve2 -verify-machineinstrs -force-streaming < %s | FileCheck %s --check-prefixes=PAIR
```
Is this different from the RUN command on line 4?

sdesmalen-arm commented on llvm/test/CodeGen/AArch64/sve-callee-save-restore-pairs.ll:
```diff
@@ -3,7 +3,8 @@
 ; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sme2 -verify-machineinstrs -force-streaming < %s | FileCheck %s --check-prefixes=PAIR
 ; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sve2p1 -verify-machineinstrs < %s | FileCheck %s --check-prefixes=PAIR
-
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sme2 -mattr=+sve2 -verify-machineinstrs < %s | FileCheck %s --check-prefixes=NOPAIR
```
Does `-mattr=+sve2` make a difference for this test? (I thought it was only the `-force-streaming`). If not, please remove it.

efriedma-quic replied:
SME without SVE should work.  And it's an increasingly important use-case, given there's hardware shipping in this configuration.

sdesmalen-arm commented:
> Is the revised check correct with __arm_locally_streaming?

It should be, the `isStreaming()` gets its value from `AArch64TargetMachine::getSubtargetImpl` which sets it if either `-force-streaming=true`, or if the function has the (locally) streaming attribute.

It's probably worth adding a test-case though.

efriedma-quic commented:
For most code, isStreaming() is the correct check, but I think the generated code here run before we switch into streaming mode.

CarolineConcatto committed:
Address review comments

This patch:

- Replace isSve2p1OrSme2InStreaming by enableMultiVectorSpillFill
- adds a aarch64-disable-multivector-spill-fill clang option
- add a test for streaming compatible function

CarolineConcatto commented:
> It should be, the isStreaming() gets its value from AArch64TargetMachine::getSubtargetImpl which sets it if either -force-streaming=true, or if the function has the (locally) streaming attribute.
> 
> It's probably worth adding a test-case though.

I believe I added a check for that, but the test for streaming compatible does not use the pair of LD/ST .
What am I missing?

CarolineConcatto committed:
Add locally streaming test for sme

efriedma-quic commented on llvm/test/CodeGen/AArch64/sme-callee-save-restore-pairs.ll:
```diff
@@ -0,0 +1,443 @@
+; PAIR-NEXT:    str p6, [sp, #13, mul vl] // 2-byte Folded Spill
+; PAIR-NEXT:    str p5, [sp, #14, mul vl] // 2-byte Folded Spill
+; PAIR-NEXT:    str p4, [sp, #15, mul vl] // 2-byte Folded Spill
+; PAIR-NEXT:    smstart sm
```
smstart is here: any SME2 instructions generated before this, including pair stores, are invalid.

CarolineConcatto replied:
So it is only safe to use sme2 instructions when in streaming mode. The compiler should avoid to use when in  streaming compatible and locally streaming . At least that is the conclusion that I have. 
I am not sure if it should check smstart and smstop when using the sme2 instructions, or if that is even possible here.

efriedma-quic replied:
For a locally-streaming function, spilling callee-saves can't happen in streaming mode: smstart will clobber the registers, so we have to spill before that.

CarolineConcatto committed:
Avoid use of sme2 instructions when in locally streaming mode

efriedma-quic reviewed (approved):
LGTM

CarolineConcatto commented:
I pushed a patch to fix the test output:
commit f627c453dbef21d6949dabbce2e7a424b9deda30
Author: Caroline Concatto <caroline.concatto@arm.com>
Date:   Mon Sep 30 08:25:13 2024 +0000

    Fix test for PR#109680

    The patch https://github.com/llvm/llvm-project/pull/109680 is failing
    because of the test sme-callee-save-restore-pairs.ll.
    This patch fixes the output of the test
